### ISSUE TITLE
boost::asio::use_future can't be compiled with libstdc++-5-dev:i386 5.3.1-11 (current Debian testing)

### ISSUE THREAD

**Issue by paulfertser** [chriskohlhoff/asio#112]
```
$ g++-5 -std=c++14 -Wall -pedantic -ggdb3 -pthread vtrackerd.cpp -lboost_system -o vtrackerd
In file included from /usr/include/c++/5/ext/alloc_traits.h:36:0,
                 from /usr/include/c++/5/bits/basic_string.h:40,
                 from /usr/include/c++/5/string:52,
                 from /usr/include/c++/5/bits/locale_classes.h:40,
                 from /usr/include/c++/5/bits/ios_base.h:41,
                 from /usr/include/c++/5/ios:42,
                 from /usr/include/c++/5/ostream:38,
                 from /usr/include/c++/5/iostream:39,
                 from vtrackerd.cpp:1:
/usr/include/c++/5/bits/alloc_traits.h: In instantiation of ‘static void std::allocator_traits<std::allocator<_CharT> >::construct(std::allocator_traits<std::allocator<_CharT> >::allocator_type&, _Up*, _Args&& ...) [with _Up = std::promise<unsigned int>; _Args = {const std::allocator_arg_t&, std::allocator<void>}; _Tp = void; std::allocator_traits<std::allocator<_CharT> >::allocator_type = std::allocator<void>]’:
/usr/include/c++/5/bits/shared_ptr_base.h:522:39:   required from ‘std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_Sp_counted_ptr_inplace(_Alloc, _Args&& ...) [with _Args = {const std::allocator_arg_t&, std::allocator<void>}; _Tp = std::promise<unsigned int>; _Alloc = std::allocator<void>; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2u]’
/usr/include/c++/5/bits/shared_ptr_base.h:617:4:   required from ‘std::__shared_count<_Lp>::__shared_count(std::_Sp_make_shared_tag, _Tp*, const _Alloc&, _Args&& ...) [with _Tp = std::promise<unsigned int>; _Alloc = std::allocator<void>; _Args = {const std::allocator_arg_t&, std::allocator<void>}; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2u]’
/usr/include/c++/5/bits/shared_ptr_base.h:1097:35:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_make_shared_tag, const _Alloc&, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {const std::allocator_arg_t&, std::allocator<void>}; _Tp = std::promise<unsigned int>; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2u]’
/usr/include/c++/5/bits/shared_ptr.h:319:64:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_make_shared_tag, const _Alloc&, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {const std::allocator_arg_t&, std::allocator<void>}; _Tp = std::promise<unsigned int>]’
/usr/include/c++/5/bits/shared_ptr.h:620:39:   required from ‘std::shared_ptr<_Tp1> std::allocate_shared(const _Alloc&, _Args&& ...) [with _Tp = std::promise<unsigned int>; _Alloc = std::allocator<void>; _Args = {const std::allocator_arg_t&, std::allocator<void>}]’
/usr/include/boost/asio/impl/use_future.hpp:39:56:   required from ‘boost::asio::detail::promise_handler<T>::promise_handler(boost::asio::use_future_t<OtherAllocator>) [with Allocator = std::allocator<void>; T = unsigned int]’
/usr/include/boost/asio/async_result.hpp:62:21:   required from ‘boost::asio::detail::async_result_init<Handler, Signature>::async_result_init(Handler&&) [with Handler = const boost::asio::use_future_t<>&; Signature = void(boost::system::error_code, unsigned int)]’
/usr/include/boost/asio/impl/read_until.hpp:711:49:   required from ‘typename boost::asio::async_result<typename boost::asio::handler_type<WriteHandler, void(boost::system::error_code, unsigned int)>::type>::type boost::asio::async_read_until(AsyncReadStream&, boost::asio::basic_streambuf<Allocator>&, const string&, ReadHandler&&) [with AsyncReadStream = boost::asio::basic_stream_socket<boost::asio::ip::tcp>; Allocator = std::allocator<char>; ReadHandler = const boost::asio::use_future_t<>&; typename boost::asio::async_result<typename boost::asio::handler_type<WriteHandler, void(boost::system::error_code, unsigned int)>::type>::type = std::future<unsigned int>; std::__cxx11::string = std::__cxx11::basic_string<char>]’
vtrackerd.cpp:46:36:   required from here
/usr/include/c++/5/bits/alloc_traits.h:530:4: error: ‘using allocator_type = class std::allocator<void> {aka class std::allocator<void>}’ has no member named ‘construct’
  { __a.construct(__p, std::forward<_Args>(__args)...); }
    ^
/usr/include/c++/5/bits/alloc_traits.h: In instantiation of ‘static void std::allocator_traits<std::allocator<_CharT> >::destroy(std::allocator_traits<std::allocator<_CharT> >::allocator_type&, _Up*) [with _Up = std::promise<unsigned int>; _Tp = void; std::allocator_traits<std::allocator<_CharT> >::allocator_type = std::allocator<void>]’:
/usr/include/c++/5/bits/shared_ptr_base.h:531:35:   required from ‘void std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_M_dispose() [with _Tp = std::promise<unsigned int>; _Alloc = std::allocator<void>; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2u]’
vtrackerd.cpp:101:1:   required from here
/usr/include/c++/5/bits/alloc_traits.h:542:4: error: ‘using allocator_type = class std::allocator<void> {aka class std::allocator<void>}’ has no member named ‘destroy’
  { __a.destroy(__p); }
    ^
```

**Comment by glenfe** [comment]
Paul's issue is due to a change in libstdc++. For example, this will also fail to compile now:
```cpp
#include <memory>
int main()
{
  auto p = std::allocate_shared<int>(std::allocator<void>());
}
```
This is because the allocator is not rebound to the `value_type` for construction and destruction. It expects the allocator instance provided to be one that is capable of constructing and destroying the `value_type` of the `shared_ptr`.

**Comment by 2underscores-vic** [comment]
The same issue when using GCC 6.1

**Comment by pjh4993** [comment]
i have same issue when i compiled boost::use_future example in official site
below is it's tutorial code
`#include <array>
#include <future>
#include <iostream>
#include <thread>
#include <boost/asio/io_service.hpp>
#include <boost/asio/ip/udp.hpp>
#include <boost/asio/use_future.hpp>
using boost::asio::ip::udp;
void get_daytime(boost::asio::io_service& ios, const char* hostname){
	try{
		udp::resolver res(ios);
		std::future<udp::resolver::iterator> iter = 
			res.async_resolve(
					{udp::v4(),hostname,"daytime"},
					boost::asio::use_future);
		udp::socket sock(ios,udp::v4());
		std::array<char,1> send_buf = {{0}};
		std::future<std::size_t> send_length = 
			sock.async_send_to(boost::asio::buffer(send_buf),
					*iter.get(),boost::asio::use_future);
		send_length.get();
		std::array<char,128> recv_buf;
		udp::endpoint sender_endpoint;
		std::future<std::size_t> recv_length = 
			sock.async_receive_from(
					boost::asio::buffer(recv_buf),
					sender_endpoint,
					boost::asio::use_future);
		std::cout.write(recv_buf.data(),recv_length.get());
	}catch(std::system_error &e){
		std::cerr << e.what() << std::endl;
	}
}
int main(int argc, char* argv[]){
	try{
		if(argc!=2){
			std::cerr << "Usage : daytime_client <host>" << std::endl;
			return 1;
		}
		boost::asio::io_service ios;
		boost::asio::io_service::work wrk(ios);
		std::thread thread([&ios](){ios.run();});
		get_daytime(ios,argv[1]);
		ios.stop();
		thread.join();
	}catch(std::exception& e){
		std::cerr << e.what() << std::endl;
	}
	return 0;
}
`
and the compile result was
`In file included from /usr/include/c++/5/ext/alloc_traits.h:36:0,
                 from /usr/include/c++/5/bits/basic_string.h:40,
                 from /usr/include/c++/5/string:52,
                 from /usr/include/c++/5/stdexcept:39,
                 from /usr/include/c++/5/array:38,
                 from daytime_client.cpp:1:
/usr/include/c++/5/bits/alloc_traits.h: In instantiation of ‘static void std::allocator_traits<std::allocator<_CharT> >::construct(std::allocator_traits<std::allocator<_CharT> >::allocator_type&, _Up*, _Args&& ...) [with _Up = std::promise<boost::asio::ip::basic_resolver_iterator<boost::asio::ip::udp> >; _Args = {const std::allocator_arg_t&, std::allocator<void>}; _Tp = void; std::allocator_traits<std::allocator<_CharT> >::allocator_type = std::allocator<void>]’:
/usr/include/c++/5/bits/shared_ptr_base.h:522:39:   required from ‘std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_Sp_counted_ptr_inplace(_Alloc, _Args&& ...) [with _Args = {const std::allocator_arg_t&, std::allocator<void>}; _Tp = std::promise<boost::asio::ip::basic_resolver_iterator<boost::asio::ip::udp> >; _Alloc = std::allocator<void>; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2u]’
/usr/include/c++/5/bits/shared_ptr_base.h:617:4:   required from ‘std::__shared_count<_Lp>::__shared_count(std::_Sp_make_shared_tag, _Tp*, const _Alloc&, _Args&& ...) [with _Tp = std::promise<boost::asio::ip::basic_resolver_iterator<boost::asio::ip::udp> >; _Alloc = std::allocator<void>; _Args = {const std::allocator_arg_t&, std::allocator<void>}; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2u]’
/usr/include/c++/5/bits/shared_ptr_base.h:1097:35:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_make_shared_tag, const _Alloc&, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {const std::allocator_arg_t&, std::allocator<void>}; _Tp = std::promise<boost::asio::ip::basic_resolver_iterator<boost::asio::ip::udp> >; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2u]’
/usr/include/c++/5/bits/shared_ptr.h:319:64:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_make_shared_tag, const _Alloc&, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {const std::allocator_arg_t&, std::allocator<void>}; _Tp = std::promise<boost::asio::ip::basic_resolver_iterator<boost::asio::ip::udp> >]’
/usr/include/c++/5/bits/shared_ptr.h:620:39:   required from ‘std::shared_ptr<_Tp1> std::allocate_shared(const _Alloc&, _Args&& ...) [with _Tp = std::promise<boost::asio::ip::basic_resolver_iterator<boost::asio::ip::udp> >; _Alloc = std::allocator<void>; _Args = {const std::allocator_arg_t&, std::allocator<void>}]’
/usr/local/include/boost/asio/impl/use_future.hpp:39:56:   required from ‘boost::asio::detail::promise_handler<T>::promise_handler(boost::asio::use_future_t<OtherAllocator>) [with Allocator = std::allocator<void>; T = boost::asio::ip::basic_resolver_iterator<boost::asio::ip::udp>]’
/usr/local/include/boost/asio/async_result.hpp:62:21:   required from ‘boost::asio::detail::async_result_init<Handler, Signature>::async_result_init(Handler&&) [with Handler = const boost::asio::use_future_t<>&; Signature = void(boost::system::error_code, boost::asio::ip::basic_resolver_iterator<boost::asio::ip::udp>)]’
/usr/local/include/boost/asio/ip/resolver_service.hpp:125:54:   required from ‘typename boost::asio::async_result<typename boost::asio::handler_type<ResolveHandler, void(boost::system::error_code, boost::asio::ip::basic_resolver_iterator<Protocol>)>::type>::type boost::asio::ip::resolver_service<InternetProtocol>::async_resolve(boost::asio::ip::resolver_service<InternetProtocol>::implementation_type&, const query_type&, ResolveHandler&&) [with ResolveHandler = const boost::asio::use_future_t<>&; InternetProtocol = boost::asio::ip::udp; typename boost::asio::async_result<typename boost::asio::handler_type<ResolveHandler, void(boost::system::error_code, boost::asio::ip::basic_resolver_iterator<Protocol>)>::type>::type = std::future<boost::asio::ip::basic_resolver_iterator<boost::asio::ip::udp> >; boost::asio::ip::resolver_service<InternetProtocol>::implementation_type = std::shared_ptr<void>; boost::asio::ip::resolver_service<InternetProtocol>::query_type = boost::asio::ip::basic_resolver_query<boost::asio::ip::udp>]’
/usr/local/include/boost/asio/ip/basic_resolver.hpp:167:54:   required from ‘typename boost::asio::async_result<typename boost::asio::handler_type<ResolveHandler, void(boost::system::error_code, boost::asio::ip::basic_resolver_iterator<Protocol>)>::type>::type boost::asio::ip::basic_resolver<InternetProtocol, ResolverService>::async_resolve(const query&, ResolveHandler&&) [with ResolveHandler = const boost::asio::use_future_t<>&; InternetProtocol = boost::asio::ip::udp; ResolverService = boost::asio::ip::resolver_service<boost::asio::ip::udp>; typename boost::asio::async_result<typename boost::asio::handler_type<ResolveHandler, void(boost::system::error_code, boost::asio::ip::basic_resolver_iterator<Protocol>)>::type>::type = std::future<boost::asio::ip::basic_resolver_iterator<boost::asio::ip::udp> >; boost::asio::ip::basic_resolver<InternetProtocol, ResolverService>::query = boost::asio::ip::basic_resolver_query<boost::asio::ip::udp>]’
daytime_client.cpp:15:29:   required from here
/usr/include/c++/5/bits/alloc_traits.h:530:4: error: ‘using allocator_type = class std::allocator<void> {aka class std::allocator<void>}’ has no member named ‘construct’
  { __a.construct(__p, std::forward<_Args>(__args)...); }
    ^
/usr/include/c++/5/bits/alloc_traits.h: In instantiation of ‘static void std::allocator_traits<std::allocator<_CharT> >::construct(std::allocator_traits<std::allocator<_CharT> >::allocator_type&, _Up*, _Args&& ...) [with _Up = std::promise<long unsigned int>; _Args = {const std::allocator_arg_t&, std::allocator<void>}; _Tp = void; std::allocator_traits<std::allocator<_CharT> >::allocator_type = std::allocator<void>]’:
/usr/include/c++/5/bits/shared_ptr_base.h:522:39:   required from ‘std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_Sp_counted_ptr_inplace(_Alloc, _Args&& ...) [with _Args = {const std::allocator_arg_t&, std::allocator<void>}; _Tp = std::promise<long unsigned int>; _Alloc = std::allocator<void>; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2u]’
/usr/include/c++/5/bits/shared_ptr_base.h:617:4:   required from ‘std::__shared_count<_Lp>::__shared_count(std::_Sp_make_shared_tag, _Tp*, const _Alloc&, _Args&& ...) [with _Tp = std::promise<long unsigned int>; _Alloc = std::allocator<void>; _Args = {const std::allocator_arg_t&, std::allocator<void>}; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2u]’
/usr/include/c++/5/bits/shared_ptr_base.h:1097:35:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_make_shared_tag, const _Alloc&, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {const std::allocator_arg_t&, std::allocator<void>}; _Tp = std::promise<long unsigned int>; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2u]’
/usr/include/c++/5/bits/shared_ptr.h:319:64:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_make_shared_tag, const _Alloc&, _Args&& ...) [with _Alloc = std::allocator<void>; _Args = {const std::allocator_arg_t&, std::allocator<void>}; _Tp = std::promise<long unsigned int>]’
/usr/include/c++/5/bits/shared_ptr.h:620:39:   required from ‘std::shared_ptr<_Tp1> std::allocate_shared(const _Alloc&, _Args&& ...) [with _Tp = std::promise<long unsigned int>; _Alloc = std::allocator<void>; _Args = {const std::allocator_arg_t&, std::allocator<void>}]’
/usr/local/include/boost/asio/impl/use_future.hpp:39:56:   required from ‘boost::asio::detail::promise_handler<T>::promise_handler(boost::asio::use_future_t<OtherAllocator>) [with Allocator = std::allocator<void>; T = long unsigned int]’
/usr/local/include/boost/asio/async_result.hpp:62:21:   required from ‘boost::asio::detail::async_result_init<Handler, Signature>::async_result_init(Handler&&) [with Handler = const boost::asio::use_future_t<>&; Signature = void(boost::system::error_code, long unsigned int)]’
/usr/local/include/boost/asio/datagram_socket_service.hpp:354:52:   required from ‘typename boost::asio::async_result<typename boost::asio::handler_type<WriteHandler, void(boost::system::error_code, long unsigned int)>::type>::type boost::asio::datagram_socket_service<Protocol>::async_send_to(boost::asio::datagram_socket_service<Protocol>::implementation_type&, const ConstBufferSequence&, const endpoint_type&, boost::asio::socket_base::message_flags, WriteHandler&&) [with ConstBufferSequence = boost::asio::mutable_buffers_1; WriteHandler = const boost::asio::use_future_t<>&; Protocol = boost::asio::ip::udp; typename boost::asio::async_result<typename boost::asio::handler_type<WriteHandler, void(boost::system::error_code, long unsigned int)>::type>::type = std::future<long unsigned int>; boost::asio::datagram_socket_service<Protocol>::implementation_type = boost::asio::detail::reactive_socket_service<boost::asio::ip::udp>::implementation_type; boost::asio::datagram_socket_service<Protocol>::endpoint_type = boost::asio::ip::basic_endpoint<boost::asio::ip::udp>; boost::asio::socket_base::message_flags = int]’
/usr/local/include/boost/asio/basic_datagram_socket.hpp:531:52:   required from ‘typename boost::asio::async_result<typename boost::asio::handler_type<WriteHandler, void(boost::system::error_code, long unsigned int)>::type>::type boost::asio::basic_datagram_socket<Protocol, DatagramSocketService>::async_send_to(const ConstBufferSequence&, const endpoint_type&, WriteHandler&&) [with ConstBufferSequence = boost::asio::mutable_buffers_1; WriteHandler = const boost::asio::use_future_t<>&; Protocol = boost::asio::ip::udp; DatagramSocketService = boost::asio::datagram_socket_service<boost::asio::ip::udp>; typename boost::asio::async_result<typename boost::asio::handler_type<WriteHandler, void(boost::system::error_code, long unsigned int)>::type>::type = std::future<long unsigned int>; boost::asio::basic_datagram_socket<Protocol, DatagramSocketService>::endpoint_type = boost::asio::ip::basic_endpoint<boost::asio::ip::udp>]’
daytime_client.cpp:20:41:   required from here
/usr/include/c++/5/bits/alloc_traits.h:530:4: error: ‘using allocator_type = class std::allocator<void> {aka class std::allocator<void>}’ has no member named ‘construct’
/usr/include/c++/5/bits/alloc_traits.h: In instantiation of ‘static void std::allocator_traits<std::allocator<_CharT> >::destroy(std::allocator_traits<std::allocator<_CharT> >::allocator_type&, _Up*) [with _Up = std::promise<long unsigned int>; _Tp = void; std::allocator_traits<std::allocator<_CharT> >::allocator_type = std::allocator<void>]’:
/usr/include/c++/5/bits/shared_ptr_base.h:531:35:   required from ‘void std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_M_dispose() [with _Tp = std::promise<long unsigned int>; _Alloc = std::allocator<void>; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2u]’
daytime_client.cpp:50:1:   required from here
/usr/include/c++/5/bits/alloc_traits.h:542:4: error: ‘using allocator_type = class std::allocator<void> {aka class std::allocator<void>}’ has no member named ‘destroy’
  { __a.destroy(__p); }
    ^
/usr/include/c++/5/bits/alloc_traits.h: In instantiation of ‘static void std::allocator_traits<std::allocator<_CharT> >::destroy(std::allocator_traits<std::allocator<_CharT> >::allocator_type&, _Up*) [with _Up = std::promise<boost::asio::ip::basic_resolver_iterator<boost::asio::ip::udp> >; _Tp = void; std::allocator_traits<std::allocator<_CharT> >::allocator_type = std::allocator<void>]’:
/usr/include/c++/5/bits/shared_ptr_base.h:531:35:   required from ‘void std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_M_dispose() [with _Tp = std::promise<boost::asio::ip::basic_resolver_iterator<boost::asio::ip::udp> >; _Alloc = std::allocator<void>; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2u]’
daytime_client.cpp:50:1:   required from here
/usr/include/c++/5/bits/alloc_traits.h:542:4: error: ‘using allocator_type = class std::allocator<void> {aka class std::allocator<void>}’ has no member named ‘destroy’
`

**Comment by pjh4993** [comment]
i have a same issue when compiling boost::use_future
i used tutorial code in boost official site
[http://www.boost.org/doc/libs/1_61_0/doc/html/boost_asio/example/cpp11/futures/daytime_client.cpp](url)
and my compile result was like this
[Uploading result.txt…]()

**Comment by caryan** [comment]
This appears to have been fixed by 5bebdf6e8cb0d87494100bdb26e3e5f8ce7e1984

**Comment by chriskohlhoff** [comment]
Was fixed on master by 5bebdf6e8cb0d87494100bdb26e3e5f8ce7e1984 and 503b8bb172240dfd3cf2b986799165c223c02a57.

Fixed on asio-1-10-branch by c116de6902f6e528a5b595b4375b87f85abea053.
